### ISSUE TITLE
[DFA-arch T1] Criteo-style CTR MLP tower + DFA — near-BP positive, stale since 2020, cheap to revive

### ISSUE THREAD

**Issue by OriPekelman** [OriPekelman/toy#154]
**Tier 1.** Second-safest positive. LightOn 2020 (arXiv:2006.12878): DFA within ~0.01 AUC of BP on Criteo (DeepFM, AFN) = effectively matching. Scalar output (CTR), huge embedding tables (want layer-parallel MLP updates), untouched since 2020 -> a modern-optimizer refresh is a clean positive on its own.

## Ask
A **CTR/recommender MLP tower**: embedding layers (categorical + numeric) -> MLP -> sigmoid CTR head + logloss/AUC, on a Criteo-style tabular set (or a compact synthetic CTR generator if data plumbing is heavy). Support per-layer `--policy chain|dfa`; DFA the MLP tower (embeddings stay chain).

## Success target
all-DFA-tower within ~0.01 AUC of all-BP. If so, a second independent positive at scalar output dim.

## Consuming Tao experiment: F18 (CTR).
Refs: LightOn 2020 (Criteo DeepFM/AFN near-parity). Note: whether DFA edges BP on DeepFM is run-dependent — treat as near-parity, not a confident win.

**Comment by OriPekelman** [comment]
Shipped in ae3796f — and **the success target is not met.** That is the result, not a build problem.

`toy train ctr`: per-field embedding tables → concat → MLP tower → **scalar** sigmoid head → logloss, per-tower-layer `--policy chain|dfa|frozen`, scored by AUC (exact Mann-Whitney, ties at 0.5 — no binning to argue about at the 0.01 scale this ticket cares about). CPU-only, `CTR_*` env namespace.

## The numbers

2000 steps, 40 teacher pairs, matched init + seed, 2048 held-out rows:

| seed | all-BP | all-DFA | frozen | BP−DFA | DFA−frozen |
|---|---|---|---|---|---|
| 0 | 0.706 | 0.639 | 0.620 | .067 | .019 |
| 1 | 0.701 | 0.643 | 0.619 | .058 | .024 |
| 2 | 0.697 | 0.638 | 0.617 | .059 | .022 |

DFA clears the frozen control reproducibly — so it *is* learning something — but lands **6 points behind BP, not 1**. LR is not the explanation: swept 0.0003 / 0.001 / 0.003 / 0.01, and 0.003 is DFA's *own* best.

## Why — and it is provable, not a guess

At output dim 1 the DFA update is **exactly rank-1**:

```
delta_l  = B_l · e = b_l ⊗ e          (a FIXED vector times a per-sample scalar)
grad W_l = b_l ⊗ ( Σᵢ eᵢ aᵢ )
```

Every policied layer can only move its output along **one fixed direction**; the rest of the layer stays at random init.

The interesting part: this is *equally true at 2 classes* (softmax CE makes the two error rows negatives of each other), and there — toy#152 — DFA **matched** BP. So rank-1 feedback is sufficient when the hidden layers only need a simple transformation, and insufficient when they must synthesise genuinely new features (here: pairwise crosses).

**That means the output-dim lens is two competing effects, not one:** fewer directions to align (helps) versus less information per step (hurts). toy#152 measured the first; this lane is where the second bites. I'd suggest updating the program's framing accordingly — "small output dim is good for DFA" is too simple.

## Why this does not contradict LightOn — and it is measurable

DeepFM, the architecture this ticket cites, is an FM branch **plus** a DNN tower, summed. I implemented the real second-order FM term (`--fm-branch`: first-order + `½Σⱼ[(Σ_f e_f[j])² − Σ_f e_f[j]²]`):

| arm | AUC |
|---|---|
| BP tower + FM | 0.806 |
| **frozen tower + FM** | **0.813** |
| DFA tower + FM | 0.651 |

**The frozen tower scores as well as the trained one.** On a DeepFM shape the tower contributes ~nothing on this task — so a "DFA within 0.01 AUC of BP on DeepFM" observation is **not evidence that DFA trains towers well**. The FM branch is doing the work. I have not read their code, so treat this as the most plausible reading rather than a proven claim about their run — but it is cheap for F18 to check, and it changes what the experiment should conclude.

Also: DFA at 0.651 is **16 points below frozen** there. That is the toy#141 shape (frozen experts beat both dfa and chain) recurring on a second independent lane — a DFA-trained module summed into a path a BP-trained component already handles is net-harmful.

## Two build notes

**The construction had to be toy#158's, not toy#152's.** This ticket says "DFA the MLP tower (embeddings stay chain)". toy#152's direct-gradient rule propagates *nothing*, so the tables below the tower would have sat at init and that requirement would be unimplementable. This lane uses a surrogate loss root per policied layer; the lowest one's gradient continues into the embeddings. The gate asserts it rather than trusting it — the frozen-tower arm still improves 0.529 → 0.620 AUC, which is only possible if the tables are learning through it.

**The task had to be built to discriminate.** My first version reproduced toy#152's blobs failure exactly: all three arms within **0.003 AUC with the frozen control ahead** (.795 frozen / .793 dfa / .792 chain), because the teacher's additive part dominated and embeddings + head alone solved it. `--lin-scale 0.25` with 40 pairs makes the crosses carry the signal. The gate asserts `BP − frozen > 0.05` **first**, so a future change that quietly makes the task easy fails loudly instead of reporting a free pass.

## What the gate does and does not assert

It pins the reproducible ordering (BP > DFA > frozen, with margins) and **prints** the BP−DFA gap. It deliberately does **not** assert the ~0.01 target: gating on a falsehood is worse than not gating. If DFA is later improved on this lane, that shows up as a number moving rather than a gate flipping.

Full write-up in `docs/roadmap/dfa-arch-program-2026-08-10.md`. 48 gate targets green (56 legs).

**Comment by OriPekelman** [comment]
Closing: shipped in `ae3796f` — feat(dfa-arch): toy#154/F18 CTR tower — and DFA does NOT reach near-parity

Verified by `git log`; the implementing commit names this issue in its subject. Part of the 2026-08-14 issue-state sweep (48 closed, all against a commit in `main`). Reopen if the scope here outran what landed.
